### PR TITLE
feat: Implement Keplr logout

### PR DESCRIPTION
**Pull request by blushi**
## Description

Closes: regen-network/regen-registry#1558

- add `useLogout` hook
- use `fetch` in `login` and `logout`
- disable logging in if using WalletConnect
- fixes some circular dependencies issues

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] provided a link to the relevant issue or specification
- [x] provided instructions on how to test
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### How to test

https://deploy-preview-1819--regen-registry.netlify.app/

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**._

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] verified React components follow DRY principles
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

### CODE REVIEW

**Review by wgwz — APPROVED** [review]
utACK

File: web-registry/src/components/organisms/RegistryLayout/RegistryLayout.Header.tsx [hunk 1]
```diff
@@ -51,7 +55,7 @@ const RegistryLayoutHeader: React.FC = () => {
         onLogin={() =>
           loginWithRedirect({ redirectUri: window.location.origin })
         }
-        onLogout={() => logout({ returnTo: window.location.origin })}
+        onLogout={() => auth0Logout({ returnTo: window.location.origin })}
```
Comment (flagrede) [inline web-registry/src/components/organisms/RegistryLayout/RegistryLayout.Header.tsx:58]: I'm confused, I thought we weren't using auth0 anymore.

**Review by flagrede — COMMENTED** [review]
I did some tests with the shared account, the keplr `Prove Ownership` modal is infinitely loading after login but it did log me in. When I try to disconnect afterward I have a `401 Unauthorized`.

**Comment by blushi** [comment]
> When I try to disconnect afterward I have a 401 Unauthorized.

This is related to the server issue I've described just above: https://github.com/regen-network/regen-web/pull/1819#issuecomment-1477450735

**Comment by flagrede** [comment]
> > When I try to disconnect afterward I have a 401 Unauthorized.
> 
> This is related to the server issue I've described just above: [#1819 (comment)](https://github.com/regen-network/regen-web/pull/1819#issuecomment-1477450735)

@blushi sorry I missed that.

**Comment by wgwz** [comment]
@blushi the only difference i can see here which looks unimportant, but might actually be resulting in some differences is the fact that the get CSRF token query uses the fetch API, whereas the login/logout request are using axios. in particular the fetch API may be handling the cookie differently than axios.

notice how the fetch parameter for `credentials` accepts a few different options (https://developer.mozilla.org/en-US/docs/Web/API/fetch#parameters). our CSRF token query is using the 'include' option for that parameter. whereas the login request and the axios parameter we are using is called 'withCredentials' and it only accepts a boolean value.

axios might actually be using the 'same-origin' method as it's documented in the fetch parameters. this could explain why the user session is not being set for deploy-previews. i briefly looked to see if axios has a way to do what the fetch API does for the "include" parameter, but didn't find anything. there might be a way though, otherwise, maybe we need to reconsider switching this to use the fetch API

**Comment by blushi** [comment]
@wgwz using `fetch` doesn't seem to fix the issue, the login response still doesn't include the Set-Cookie header at all, which makes me think it might be server related... still investigating

**Comment by blushi** [comment]
@flagrede Kyle found out the issue with the /login endpoint which was server-related so you can test it again.
It wasn't related to using `axios` over `fetch` at the end but I did migrate `useLogin` to use `fetch` instead, and you were suggesting to do this in my initial Keplr login PR. This is a bit more verbose now without any util wrapping functions set up, what do you think? should I migrate `useLogout` too or just rollback to the axios-based implementation?

**Comment by flagrede** [comment]
> @flagrede Kyle found out the issue with the /login endpoint which was server-related so you can test it again. It wasn't related to using `axios` `withCredentials` over `fetch` `credentials` at the end but I did migrate `useLogin` to use `fetch` instead, and you were suggesting to do this in my initial Keplr login PR. This is a bit more verbose now without any util wrapping functions set up, what do you think? should I migrate `useLogout` too or just rollback to the axios-based implementation?

@blushi Thanks for the update, even if it's a bit more verbose I do prefer the fetch implementation as it's more explicit (and given our low usage I would still like to remove the lib eventually).

**Comment by blushi** [comment]
@erikalogie you can test keplr login/logout on https://deploy-preview-1819--regen-registry.netlify.app/

**Comment by blushi** [comment]
> > @flagrede Kyle found out the issue with the /login endpoint which was server-related so you can test it again. It wasn't related to using `axios` `withCredentials` over `fetch` `credentials` at the end but I did migrate `useLogin` to use `fetch` instead, and you were suggesting to do this in my initial Keplr login PR. This is a bit more verbose now without any util wrapping functions set up, what do you think? should I migrate `useLogout` too or just rollback to the axios-based implementation?
> 
> @blushi Thanks for the update, even if it's a bit more verbose I do prefer the fetch implementation as it's more explicit (and given our low usage I would still like to remove the lib eventually).

@flagrede SGTM, done

**Comment by erikalogie** [comment]
@blushi 
Do you know why this is happening? https://monosnap.com/file/WNJYY6318qP1VsBcm0DAb8jWdGiaws

Also, I can't connect with wallet connect on this deploy preview, is that expected?

This project card also looks strange: https://monosnap.com/file/w9YhUgcb1heVr6nFimwSFuYwISOJ6W

**Comment by erikalogie** [comment]
Also, I'm noticing that I can log out, log in and the close the keplr transaction signing window, and still see all the things that one should see when logged in, even thought it still shows the "log in" button on the top right. Then if I actually go to initiate a transaction I get an error: https://monosnap.com/file/z3Fc3rSCSRWTNEh2iVzuTtglrv061i

Seems like we should not actually log them in and allow them to see the content until after they sign the transaction?

**Review by flagrede — APPROVED** [review]
Looks good!

**Comment by blushi** [comment]
> Do you know why this is happening? https://monosnap.com/file/WNJYY6318qP1VsBcm0DAb8jWdGiaws
> 
> Also, I can't connect with wallet connect on this deploy preview, is that expected?

looking into this, seems like there's an issue with wallet connect indeed

> This project card also looks strange: https://monosnap.com/file/w9YhUgcb1heVr6nFimwSFuYwISOJ6W

this is not related to the changes here AFAIK, best to open a bug for that

> Also, I'm noticing that I can log out, log in and the close the keplr transaction signing window, and still see all the things that one should see when logged in, even thought it still shows the "log in" button on the top right. Then if I actually go to initiate a transaction I get an error: https://monosnap.com/file/z3Fc3rSCSRWTNEh2iVzuTtglrv061i
> 
> Seems like we should not actually log them in and allow them to see the content until after they sign the transaction?

good catch, just pushed a fix for that

**Comment by erikalogie** [comment]
> > This project card also looks strange: https://monosnap.com/file/w9YhUgcb1heVr6nFimwSFuYwISOJ6W
> 
> this is not related to the changes here AFAIK, best to open a bug for that

Ok, I made a task: https://app.zenhub.com/workspaces/regen-registry-5f8998bec8958d000f4609e2/issues/gh/regen-network/regen-registry/1628

**Comment by erikalogie** [comment]
> > Do you know why this is happening? https://monosnap.com/file/WNJYY6318qP1VsBcm0DAb8jWdGiaws
> > Also, I can't connect with wallet connect on this deploy preview, is that expected?
> 
> looking into this, seems like there's an issue with wallet connect indeed

Ok, let me know when it is fixed and I will re-test

**Review by flagrede — APPROVED** [review]
tACK (tested desktop and WC)

**Comment by erikalogie** [comment]
@blushi LGTM!

**Pull request merged**